fe/fe-core/src/main/java/org/apache/doris/datasource/DataSourceMgr.java

```diff
@@ -52,103 +66,180 @@ public DataSourceMgr() {
+            if (nameToDataSource.containsKey(log.getDsName())) {
+                throw new DdlException("Datasource had already exist with name: " + log.getDsName());
+            }
+            DataSourceIf ds = DataSourceFactory.constructorFromLog(log);
+            nameToDataSource.put(ds.getName(), ds);
+        } finally {
+            writeUnlock();
         }
-        out.writeInt(size);
-        for (DataSourceIf ds : idToDataSource.values()) {
-            if (ds.getId() == InternalDataSource.INTERNAL_DS_ID) {
-                continue;
+    }
+
+    /**
+     * Reply for drop ds event.
+     */
+    public void replyDropDs(DataSourceLog log) throws UserException {
+        writeLock();
+        try {
+            if (!nameToDataSource.containsKey(log.getDsName())) {
+                throw new DdlException("No datasource found with name: " + log.getDsName());
             }
-            ExternalDataSource extDs = (ExternalDataSource) ds;
-            extDs.write(out);
+            nameToDataSource.remove(log.getDsName());
+        } finally {
+            writeUnlock();
         }
-        dsMgrProperty.write(out);
     }
 
     /**
-     * read from image.
-     *
-     * @param in
-     * @return
-     * @throws IOException
+     * Reply for alter ds name event.
      */
-    public static DataSourceMgr read(DataInput in) throws IOException {
-        if (Config.disable_cluster_feature) {
-            return null;
+    public void replyAlterDsName(DataSourceLog log) throws UserException {
+        writeLock();
+        try {
+            if (!nameToDataSource.containsKey(log.getDsName())) {
+                throw new DdlException("No datasource found with name: " + log.getDsName());
+            }
+            DataSourceIf ds = nameToDataSource.remove(log.getDsName());
+            ds.modifyDatasourceName(log.getNewDsName());
+            nameToDataSource.put(ds.getName(), ds);
+        } finally {
+            writeUnlock();
         }
-        DataSourceMgr mgr = new DataSourceMgr();
-        mgr.readFields(in);
-        return mgr;
     }
 
-    private void readFields(DataInput in) throws IOException {
-        int size = in.readInt();
-        for (int i = 0; i < size; ++i) {
-            ExternalDataSource extDs = ExternalDataSource.read(in);
-            idToDataSource.put(extDs.getId(), extDs);
-            nameToDataSource.put(extDs.getName(), extDs);
+    /**
+     * Reply for alter ds props event.
+     */
+    public void replyAlterDsProps(DataSourceLog log) throws UserException {
+        writeLock();
+        try {
+            if (!nameToDataSource.containsKey(log.getDsName())) {
+                throw new DdlException("No datasource found with name: " + log.getDsName());
+            }
+            if (!nameToDataSource.get(log.getDsName()).getType().equalsIgnoreCase(log.getNewProps().get("type"))) {
+                throw new DdlException("Can't modify the type of datasource property with name: " + log.getDsName());
```

Comment:
I think we can analyze this error in `analyze()` of AlterDsStmt?

Reply:
`analyze` can't get the type of this datasource. But I will move it out of replay Funciton